core/templates/dev/head/pages/exploration_player/ImagePreloaderService.js

```diff
@@ -0,0 +1,197 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to preload image into AssetsBackendApiService's cache.
+ */
+
+oppia.factory('ImagePreloaderService', [
+  '$q', '$uibModal', 'ExplorationContextService', 'AssetsBackendApiService',
+  'ExplorationPlayerStateService', 'UrlInterpolationService',
+  'ComputeGraphService', 'ExtractImageFilenamesFromStateService',
+  function($q, $uibModal, ExplorationContextService, AssetsBackendApiService,
+      ExplorationPlayerStateService, UrlInterpolationService,
+      ComputeGraphService, ExtractImageFilenamesFromStateService) {
+    var MAX_NUM_IMAGE_FILES_TO_DOWNLOAD_SIMULTANEOUSLY = 3;
+
+    var _filenamesOfImageCurrentlyDownloading = [];
```

Comment:
In AssetsBackendApiService.js there is _imageFilesCurrentlyBeingRequested, can't we use it instead of this array? For example on line 93 you need to remove the downloaded file from this array, if we used _imageFilesCurrentlyBeingRequested this would be handled by AssetsBackendApiService.

Reply:
Yes, we can use it and that way deleting the file will be handled by it. I created another array to create similarity between the AudioPreloaderService and the ImagePreloaderService. Also, keeping it separate keeps the code more clean I guess.
What do you say ?

Reply:
Sorry. I just checked it now, _imageFilesCurrentlyBeingRequested is the array of the form FileDownloadRequestObjectFactory. So it can't be used in place of the _filenamesOfImageCurrentlyDownloading.